### PR TITLE
sdpa: replace the experimental torch op with cudnn::sdpa_fwd / cudnn::sdpa_bwd

### PR DESCRIPTION
**Pull request by vedaanta**
## Summary

Replaces the experimental SDPA torch op with the canonical `cudnn::sdpa_fwd` / `cudnn::sdpa_bwd` custom ops. Supersedes #517, which is closed — its commits are the first two here, so this PR is the whole story in one place: add the new ops, then delete the module they replace.

`cudnn.experimental.ops.sdpa` registers its own `cudnn::sdpa` plus a backward that had to be renamed `cudnn::sdpa_bwd_legacy` purely to avoid colliding with the canonical name. Removing the module retires the collision and the rename with it.

## What lands

**New: `python/cudnn/sdpa/fwd/torch_op.py`** — `cudnn::sdpa_fwd` / `cudnn::sdpa_bwd`, exposing the cuDNN SDPA surface that `F.scaled_dot_product_attention`'s aten contract cannot express: attention sinks, sliding window, bottom-right causal, padded batches, and THD/varlen `(T, H, D)` + `cu_seqlens`. Both ship `register_fake` meta kernels; the forward passes `torch.library.opcheck` including dynamic-shape AOT dispatch.

**Removed**
- `python/cudnn/experimental/ops/sdpa.py` (914 lines) and its `scaled_dot_product_attention` export. The *package* stays — it still hosts the lazy `moe_grouped_matmul` / `swiglu_mlp` aliases.
- `test/python/test_cudnn_sdpa_op.py` (593 lines, 16 tests).
- The "SDPA PyTorch Custom Op (Experimental)" section of `docs/operations/Attention.md`, the README bullet, and the stale `sdpa.py` example reference in `docs/adding_torch_custom_ops.md`.

## Review feedback (@Anerudhan)

- **Docs folded, not split across directories.** `docs/fe-oss-apis/sdpa-torch-ops.md` is gone; its content is now the *SDPA PyTorch Custom Ops* section of `docs/operations/Attention.md`, next to the FP16/BF16 and FP8 sections. Nothing linked the old file.
- **Tests modularized by op**: `test/python/sdpa/test_torch_ops.py`. One deviation from the literal request, called out because it is a silent-failure trap: `pytest.ini` sets no `python_files` override, so collection is the default `test_*.py`, and `test/python/sdpa/` is deliberately the *helpers* package (`fp16.py`, `helpers.py`, `random_config.py` are imported, never collected) with every real test under it named `test_*.py`. A file named exactly `torch_ops.py` would have stopped being collected and all 19 tests would have gone quiet. Verified 19 collected both directly and in the default full sweep.
- **JAX equivalent**: parked for now per your note.

Also folds in a CodeRabbit finding on the same code: `_check_same_device` was receiving only the auxiliary operands, so a `k`/`v`/`o` on a different CUDA device reached cuDNN as a foreign pointer and faulted with an illegal memory access instead of the clear error. Both call sites now pass the io tensors.

## Benchmarks

Three e2e scripts used the experimental op and now call `torch.ops.cudnn.sdpa_fwd`:

| script | needs | covered |
|---|---|---|
| `Qwen-Image/run_model.py` | padded `seq_len_q` / `seq_len_kv` | ✅ |
| `Qwen3.8/run_model.py` | full causal window | ✅ (`window_size` defaults to `(-1, -1)`; its torch A/B arm already rejects anything else) |
| `Qwen3.8/run_matrix.py` | module provenance record | ✅ |

Both arms are inference-only (`requires_grad_(False)`, no `.backward()`), so the forward-only contract suffices. They now **raise explicitly** on `dropout_p`, and Qwen3.8 on a right window bound, rather than silently ignoring either — `sdpa_fwd` has no `window_right` yet.

## Verification

- `test/python/sdpa/test_torch_ops.py`: **19 passed**.
- `cudnn` and `cudnn.experimental.ops` import cleanly; `__all__` is `['moe_grouped_matmul', 'swiglu_mlp']`, lazy aliases resolve, `scaled_dot_product_attention` raises `ImportError`.
- All three benchmark scripts compile.
- Full `test/python` collection clean, no import errors.
- `grep` for `experimental.ops.sdpa` / `cudnn_sdpa_module` across `python/ test/ docs/ benchmark/ README.md`: no hits.
- black clean across `test/ python/ benchmark/`.

Rebased on `develop` `606e16f97`.

Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>

## Summary by CodeRabbit

- **New Features**
  - Added a public cuDNN scaled dot-product attention interface for dense and variable-length attention.
  - Added support for attention sinks, sliding windows, causal alignment, padding, grouped-query attention, LSE output, and autograd integration.
  - Added public forward and backward attention operators.

- **Documentation**
  - Updated attention documentation with API usage, supported layouts, requirements, and limitations.
  - Added HSTU Attention documentation.

- **Refactor**
  - Replaced the experimental attention interface with the public API in benchmarks and examples.

### CODE REVIEW

File: docs/operations/Attention.md [hunk 1]
```diff
@@ -710,95 +710,6 @@ of selected key/value blocks. It is a standalone Python FE OSS API and is
 separate from the cuDNN Graph API described above.
```
Comment (Anerudhan) [inline docs/operations/Attention.md:794]: Why this diff?

**Comment by Anerudhan** [comment]
1. Rather than adding docs/fe-oss-apis/sdpa-torch-ops.md, can we fold link into docs/operations/Attention.md? The torch ops are another entry point to the same op, and I'd prefer not to split attention docs across directories.

2. Can we rename` test/python/test_cudnn_sdpa_torch_ops.py `to `test/python/sdpa/torch_ops.py`? Lets modularizing tests by op instead of one flat file per feature.

3. Is there a JAX equivalent of torch.ops.cudnn.sdpa_fwd ? Not blocking, but I want to make sure the API surface here doesn't box us in.

**Review by Anerudhan — DISMISSED** [review]

File: docs/operations/Attention.md (hunk 1 above)
Comment (vedaanta) [reply docs/operations/Attention.md:794]: this is because we are removing the previous experimental torch op in favour of new ones.

**Comment by vedaanta** [comment]
Thanks — all three addressed in `ce82916`, and #517 is closed so this PR is now the whole story (add the ops, remove the module they replace). Retitled accordingly.

**1. Docs folded.** `docs/fe-oss-apis/sdpa-torch-ops.md` is deleted; its content is now the *SDPA PyTorch Custom Ops* section of `docs/operations/Attention.md`, sitting with the FP16/BF16 and FP8 sections under `## API` with its own anchor. Agreed on the reasoning — it is another entry point to the same op, not a separate API. Nothing linked the old file, so there are no dangling references.

**2. Tests moved** to `test/python/sdpa/test_torch_ops.py`.

One deviation from the literal path you asked for, which I want to flag rather than bury, because it fails silently: `pytest.ini` sets no `python_files` override, so collection is the default `test_*.py`. `test/python/sdpa/` is deliberately the *helpers* package — `fp16.py`, `helpers.py`, `random_config.py` are imported by the harness and never collected — and every actual test under that tree is `test_*.py` (`sdpa/frost/test_*.py`). A file named exactly `torch_ops.py` would have been treated as a helper and quietly dropped from collection: all 19 tests would have stopped running with a green suite. The `test_` prefix keeps your structure (modularized by op under `sdpa/`) while staying collectable. Verified 19 tests both when targeting the file directly and in the default full-directory sweep.

Happy to use the bare name instead if you would rather add `python_files` to `pytest.ini`, but that changes collection repo-wide, so I left it alone.

**3. JAX** — parked for now per your note. For the record, the answer today is no: there is a `cudnn.jax` entry point for BSA, but nothing equivalent for the SDPA op. Worth its own issue if you want the API surfaces to match.

Also folded in a CodeRabbit finding on the same code while I was here: `_check_same_device` was only receiving the auxiliary operands, so a `k`/`v`/`o` on a different CUDA device would reach cuDNN as a foreign device pointer and fault with an illegal memory access instead of the helper's clear error. Both call sites now pass the io tensors.

Rebased on `develop` `606e16f97`. 19 passed, black clean.

File: docs/operations/Attention.md (hunk 1 above)
Comment (vedaanta) [reply docs/operations/Attention.md:794]: That diff removes the old **SDPA PyTorch Custom Op (Experimental)** section — the docs for `cudnn.experimental.ops.sdpa`, which this PR deletes. It was describing the module being replaced, so leaving it would have documented a module that no longer exists.

Since your other comment, the same file also *gains* a section: the folded-in `docs/fe-oss-apis/sdpa-torch-ops.md` content is now **SDPA PyTorch Custom Ops (`cudnn::sdpa_fwd` / `cudnn::sdpa_bwd`)**, in the same `## API` region the removed one occupied. So the net effect on Attention.md is a swap: the experimental op's docs out, the canonical ops' docs in, and no attention docs living outside `docs/operations/` anymore.

**Comment by vedaanta** [comment]
@cudnn-ci-bot run frost,oss

**Review by Anerudhan — APPROVED** [review]
Please address code-review comments

File: benchmark/e2e/Qwen3.8/run_model.py [hunk 2]
```diff
@@ -116,16 +116,18 @@ def _cudnn_sdpa_flash(
...
-        o = cudnn_sdpa(
+        if dropout_p:
+            raise NotImplementedError("cudnn::sdpa_fwd does not serve dropout")
+        if window_size[1] not in (-1, 0):
+            raise NotImplementedError(f"cudnn::sdpa_fwd has no right window bound; got {window_size[1]}")
+        o, _ = torch.ops.cudnn.sdpa_fwd(
             qt,
             kt,
             vt,
+            softmax_scale,
```
[the remark this answers is not in the document]
Comment (vedaanta) [reply benchmark/e2e/Qwen3.8/run_model.py]: Good catch, and it was a real break — fixed in `83e135b7`.

The adapters keep torch's `scale=None` convention ("use the default"), but `cudnn::sdpa_fwd` takes a required `float attn_scale`, so the default path would have raised at dispatch. The old experimental wrapper accepted `None` and defaulted internally, so this was introduced when the benchmarks moved over. Both sites now resolve first:

```python
attn_scale = softmax_scale if softmax_scale is not None else qt.shape[-1] ** -0.5
```

Same in `Qwen-Image/run_model.py` for its `scale` parameter.

File: test/python/sdpa/test_torch_ops.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+import torch
+
+import cudnn
+
+if not torch.cuda.is_available():
+    pytest.skip("CUDA device required", allow_module_level=True)
+if torch.cuda.get_device_capability()[0] < 8:
+    pytest.skip("cuDNN SDPA requires sm80+", allow_module_level=True)
+if cudnn.backend_version() < 90600:
+    pytest.skip("requires cuDNN >= 9.6 (THD token-major stats)", allow_module_level=True)
```
[the remark this answers is not in the document]
Comment (vedaanta) [reply test/python/sdpa/test_torch_ops.py:28]: Confirmed against the header and fixed in `83e135b7`.

`scaled_dot_product_flash_attention.h:316` rejects it outright:

```cpp
RETURN_CUDNN_FRONTEND_ERROR_IF(detail::get_backend_version() < 91300 &&
                                   attributes.inputs.find(input_names::SINK_TOKEN) != attributes.inputs.end(),
                               error_code_t::ATTRIBUTE_NOT_SET,
                               "SDPA with sink_token is not supported before 9.13.");
```

The module gate only needs 9.6 (token-major THD stats), so on a 9.6–9.12 backend those two tests would have failed rather than skipped. Added a shared marker next to the existing gates and applied it to `test_sinks` and `test_sinks_with_window`:

```python
_SINKS_UNSUPPORTED = pytest.mark.skipif(cudnn.backend_version() < 91300, reason="sink_token requires cuDNN >= 9.13")
```

File: docs/operations/Attention.md [hunk 4]
```diff
@@ -720,94 +720,76 @@ forward and backward automatically build private block metadata on the active
...
+                                  return_lse=True)
+o.backward(grad)  # routes through cudnn::sdpa_bwd via register_autograd
 
-# Backward (autograd handles this automatically)
-loss = output.sum()
-loss.backward()
-# q.grad, k.grad, v.grad are now populated
+# Or through the python wrapper (same op underneath):
+o = cudnn.sdpa_torch(q, k, v, is_causal=True, cu_seqlens_q=cu, cu_seqlens_kv=cu,
+                     max_seqlen_q=mx, max_seqlen_kv=mx)
```
[the remark this answers is not in the document]
Comment (vedaanta) [reply docs/operations/Attention.md]: Right — fixed in `83e135b7`. The wrapper example sat in the differentiable THD block but relied on the default `return_lse=False`, which the THD backward rejects. It now asks for the stats explicitly:

```python
o, lse = cudnn.sdpa_torch(q, k, v, is_causal=True, cu_seqlens_q=cu, cu_seqlens_kv=cu,
                          max_seqlen_q=mx, max_seqlen_kv=mx, return_lse=True)
```

with a comment noting the default, so the contrast with the op-level example just above it is explicit.

File: python/cudnn/sdpa/fwd/torch_op.py [hunk 5]
```diff
@@ -0,0 +1,973 @@
...
+            raise ValueError("varlen path needs cu_seqlens_q, cu_seqlens_kv, max_seqlen_q, max_seqlen_kv")
+        if q.ndim != 3:
+            raise ValueError(f"varlen path expects packed (T, H, D) tensors, got q.ndim={q.ndim}")
+        if seq_len_q is not None or seq_len_kv is not None:
+            raise ValueError("varlen path derives seq lens from cu_seqlens; do not pass seq_len_q/kv")
+        B = cu_seqlens_q.numel() - 1
+        q = _normalize_thd(q, "q")
+        k = _normalize_thd(k, "k")
+        v = _normalize_thd(v, "v")
+        T_q, H_q, D_qk = q.shape
```
[the remark this answers is not in the document]
Comment (vedaanta) [reply python/cudnn/sdpa/fwd/torch_op.py:349]: Fixed in `83e135b7` — used `T_q` rather than renaming it to `_T_q`, since it is unpacked from the same normalized `q` two lines earlier and both uses are inside the same `is_thd` branch:

```python
o = torch.empty(T_q, H_q, D_v, ...)
stats = torch.empty(T_q, H_q, 1, ...) if return_lse else ...
```
